File: ios/NHSOnline/NHSOnline/Services/KnownServices.swift
import UIKit
import SafariServices
import WebKit
import os.log
import Foundation

class KnownServices {
    let rootServices: [RootService]?
    let wildCard = "*"

    init() {
        self.rootServices = nil
    }

    init(_ rootServices: [RootService]?) {
        self.rootServices = rootServices
    }

    public func findMatchingKnownService(_ url: URL?) -> KnownService {
        guard let theUrl = url, let rootService = getRootServiceByHostAndScheme(host: (theUrl.host)!, scheme: (theUrl.scheme)!) else {
            return RootService(url: "", javaScriptInteractionMode: .None, menuTab: .None, integrationLevel: .Bronze, validateSession: false, subServices: nil)
        }
        guard let subService = findMatchingSubService(rootService: rootService, path: theUrl.path, query: theUrl.query) else {
            return rootService
        }
        return subService
    }

    private func findMatchingSubService(rootService: RootService, path: String?, query: String?) -> KnownService? {
        guard let subServices = rootService.subServices else {
            return nil
        }
        if (path == nil || path == "" || path == "/") && (query == nil) {
            return nil
        }

        var bestServiceMatch: SubServiceMatch?

        let queryItems = queryStringToArray(query)

        for subService in subServices {
            bestServiceMatch = compareServices(path: path, queryItems: queryItems, bestServiceMatch: bestServiceMatch, subService: subService)
        }
        return bestServiceMatch?.subService
    }

    private func findMatch(path: String?, queryItems: [String], subService: SubService) -> SubServiceMatch {
        let thePath = makePathSafe(path: path)
        let subServicePath = makePathSafe(path: subService.path)

        let pathMatch = subServicePath != "/" ? isPathMatch(subServicePath: subServicePath, path: thePath) : subServicePath == thePath
        let subServiceQueryItems = queryStringToArray(subService.queryString)

        if subServiceQueryItems.isEmpty {
            return SubServiceMatch.init(
                    subService: subService,
                    pathMatch: pathMatch,
                    pathMatchCount: subServicePath.count,
                    queryMatch: queryItems.isEmpty ? true : false,
                    queryMatchCount: 0)
        }

        return SubServiceMatch.init(
                subService: subService,
                pathMatch: pathMatch,
                pathMatchCount: subServicePath.count,
                queryMatch: Set(subServiceQueryItems).isSubset(of: Set(queryItems)),
                queryMatchCount: Set(subServiceQueryItems).intersection(Set(queryItems)).count
        )
    }
    
    private func isPathMatch(subServicePath: String, path: String) -> Bool {
        let pathComponents = path.split(separator: "/", omittingEmptySubsequences: true)
        let subServicePathComponents = subServicePath.split(separator: "/", omittingEmptySubsequences: true)
        
        if (pathComponents.count < subServicePathComponents.count) { return false }
        
        for i in 0 ..< subServicePathComponents.count {
            if((subServicePathComponents[i] != wildCard && subServicePathComponents[i] != pathComponents[i]) ||
                subServicePathComponents[i] == wildCard && pathComponents[i].isEmpty) {
            return false
            }
        }
        
        return true
    }

    private func makePathSafe(path: String?) -> String {
        path != nil && path != "/" ? path! + "/" : "/"
    }

    private func compareServices(path: String?, queryItems: [String], bestServiceMatch: SubServiceMatch?, subService: SubService) -> SubServiceMatch? {
        let subServiceMatch = findMatch(path: path, queryItems: queryItems, subService: subService)

        if !(subServiceMatch.pathMatch || subServiceMatch.queryMatch) {
            return bestServiceMatch
        }

        if bestServiceMatch == nil {
            return subServiceMatch.pathMatch && subServiceMatch.subService.queryString == nil || (subServiceMatch.queryMatch && subServiceMatch.queryMatchCount > 0) ? subServiceMatch : nil
        }

        if !subServiceMatch.pathMatch && bestServiceMatch!.pathMatch {
            return bestServiceMatch
        }

        if subServiceMatch.pathMatch && !bestServiceMatch!.pathMatch {
            return subServiceMatch
        }

        if !subServiceMatch.queryMatch && bestServiceMatch!.queryMatch {
            return bestServiceMatch
        }

        if subServiceMatch.queryMatch && !bestServiceMatch!.queryMatch {
            return subServiceMatch
        }

        if subServiceMatch.queryMatchCount > 0 || bestServiceMatch!.queryMatchCount > 0 {
            return subServiceMatch.queryMatchCount > bestServiceMatch!.queryMatchCount ? subServiceMatch : bestServiceMatch
        }

        return subServiceMatch.pathMatchCount > bestServiceMatch!.pathMatchCount ? subServiceMatch : bestServiceMatch
    }

    private func queryStringToArray(_ query: String?) -> [String] {
        guard let theQuery: String = query else {
            return []
        }
        return theQuery.components(separatedBy: "&")
    }

    public func getRootServiceByHostAndScheme(host: String, scheme: String) -> RootService? {
        if let i = rootServices?.firstIndex(where: {
            URL(string: $0.url)?.host! == host && URL(string: $0.url)?.scheme! == scheme
        }) {
            return rootServices?[i]
        }
        return nil
    }

    struct SubServiceMatch {
        let subService: SubService
        let pathMatch: Bool
        let pathMatchCount: Int
        let queryMatch: Bool
        let queryMatchCount: Int
    }
}

File: android/app/src/main/java/com/nhs/online/nhsonline/services/knownservices/KnownServices.kt
package com.nhs.online.nhsonline.services.knownservices

import com.nhs.online.nhsonline.services.knownservices.enums.ViewMode
import java.net.URL

data class KnownServices(
    var knownServices: List<RootService>? = null
) {
    fun findMatchingKnownService(url: URL?): KnownService? {
        var rootService: RootService?
        var subService: SubService?

        url?.let { theUrl ->
            rootService = getRootServiceByHostAndScheme(theUrl.host, theUrl.protocol)
            rootService?.let { service ->
                subService = findMatchingSubService(service, theUrl.path, theUrl.query)
                subService?.let { return it }

                return service
            }
        }

        return null
    }

    fun getRootServiceByHostAndScheme(host: String, scheme: String): RootService? {
        var i = knownServices?.indexOfFirst { service ->
            val url = URL(service.url)
            url.host == host && url.protocol == scheme
        }

        i?.let { index ->
            if (index >= 0) return knownServices!![index]
        }

        return null
    }

    fun findMatchingSubService(rootService: RootService, path: String?, query: String?): SubService? {
        if (rootService.subServices.isNullOrEmpty()) {
            return null
        }

        if ((path.isNullOrBlank() || path == "/") && query.isNullOrBlank()) {
            return null
        }

        var bestServiceMatch: SubServiceMatch? = null

        var queryItems = queryStringToArray(query)

        rootService.subServices?.forEach { subService ->
            bestServiceMatch = compareServices(path, queryItems, bestServiceMatch, subService)
        }

        return bestServiceMatch?.subService
    }

    data class SubServiceMatch (
            var subService: SubService,
            var pathMatch: Boolean,
            var pathMatchCount: Int,
            var queryMatch: Boolean,
            var queryMatchCount: Int
    )

    fun queryStringToArray(query: String?): List<String> {
        if (query.isNullOrBlank()) {
            return emptyList()
        }

        return query.split("&")
    }

    fun compareServices(path: String?, queryItems: List<String>, bestServiceMatch: SubServiceMatch?, subService: SubService): SubServiceMatch? {
        var subServiceMatch = findMatch(path, queryItems, subService)

        if (!(subServiceMatch.pathMatch || subServiceMatch.queryMatch)) {
            return bestServiceMatch
        }

        if (bestServiceMatch == null) {
            return when((subServiceMatch.pathMatch && subServiceMatch.subService.queryString.isNullOrBlank())
                    || (subServiceMatch.queryMatch && subServiceMatch.queryMatchCount > 0)) {
                true -> subServiceMatch
                false -> null
            }
        }

        if (!subServiceMatch.pathMatch && bestServiceMatch.pathMatch) {
            return bestServiceMatch
        }

        if (subServiceMatch.pathMatch && !bestServiceMatch.pathMatch) {
            return subServiceMatch
        }

        if (!subServiceMatch.queryMatch && bestServiceMatch.queryMatch) {
            return bestServiceMatch
        }

        if (subServiceMatch.queryMatch && !bestServiceMatch.queryMatch) {
            return subServiceMatch
        }

        if (subServiceMatch.queryMatchCount > 0 || bestServiceMatch.queryMatchCount > 0) {
            return when(subServiceMatch.queryMatchCount > bestServiceMatch.queryMatchCount) {
                true -> subServiceMatch
                false -> bestServiceMatch
            }
        }

        return when(subServiceMatch.pathMatchCount > bestServiceMatch.pathMatchCount) {
            true -> subServiceMatch
            false -> bestServiceMatch
        }
    }

    fun findMatch(path: String?, queryItems: List<String>, subService: SubService): SubServiceMatch {
        val thePath = when(path.isNullOrBlank() || path == "/") {
            true -> "/"
            false -> "$path/"
        }

        val subServicePath = when(subService.path.isNullOrBlank() || subService.path == "/") {
            true -> "/"
            false -> "${subService.path}/"
        }

        val pathMatch = when(subServicePath != "/") {
            true -> thePath.startsWith(subServicePath)
            false -> subServicePath == thePath
        }

        val subServiceQueryItems = queryStringToArray(subService.queryString)

        if (subServiceQueryItems.isEmpty()) {
            return SubServiceMatch(
                    subService,
                    pathMatch,
                    subServicePath.length,
                    queryItems.isEmpty(),
                    0
            )
        }

        return SubServiceMatch(
                subService,
                pathMatch,
                subServicePath.length,
                queryItems.containsAll(subServiceQueryItems),
                queryItems.intersect(subServiceQueryItems).size
        )
    }
}